File: 2026/08-August/28-Friday/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string lexPalindromicPermutation(string s, string target) {
        int n = s.size();
        vector<int> cnt(26, 0);

        for (char c : s)
            cnt[c - 'a']++;

        // A palindrome can have at most one character
        // with an odd frequency.
        int odd = 0;
        char mid = 0;

        for (int i = 0; i < 26; i++) {
            if (cnt[i] % 2) {
                odd++;
                mid = char('a' + i);
            }
        }

        if (odd > 1)
            return "";

        // Counts available for the left half.
        vector<int> half(26);
        for (int i = 0; i < 26; i++)
            half[i] = cnt[i] / 2;

        string left;

        auto buildPalindrome = [&](const string& l) {
            string res = l;

            if (mid)
                res += mid;

            for (int i = (int)l.size() - 1; i >= 0; i--)
                res += l[i];

            return res;
        };

        for (int pos = 0; pos < n / 2; pos++) {

            bool found = false;

            // Try the smallest possible character.
            for (int c = 0; c < 26; c++) {
                if (half[c] == 0)
                    continue;

                half[c]--;
                left.push_back(char('a' + c));

                // Put the largest possible characters
                // in all remaining positions.
                string remaining;

                for (int x = 25; x >= 0; x--) {
                    remaining += string(half[x], char('a' + x));
                }

                string candidateLeft = left + remaining;
                string candidate = buildPalindrome(candidateLeft);

                if (candidate > target) {
                    found = true;
                    break;
                }

                // This character cannot lead to a valid answer.
                left.pop_back();
                half[c]++;
            }

            if (!found)
                return "";
        }

        string ans = buildPalindrome(left);

        return ans > target ? ans : "";
    }
};

File: 2026/08-August/28-Friday/solution.cs
using System;

public class Solution
{
    public string LexPalindromicPermutation(string s, string target)
    {
        int n = s.Length;
        int[] count = new int[26];

        foreach (char ch in s)
            count[ch - 'a']++;

        // A palindrome can have at most one odd-frequency character.
        int odd = 0;
        char middle = '\0';

        for (int i = 0; i < 26; i++)
        {
            if (count[i] % 2 == 1)
            {
                odd++;
                middle = (char)('a' + i);
            }
        }

        if (odd > 1)
            return "";

        // Characters available for the left half.
        int[] half = new int[26];

        for (int i = 0; i < 26; i++)
            half[i] = count[i] / 2;

        string left = "";

        string BuildPalindrome(string l)
        {
            string result = l;

            if (middle != '\0')
                result += middle;

            for (int i = l.Length - 1; i >= 0; i--)
                result += l[i];

            return result;
        }

        for (int pos = 0; pos < n / 2; pos++)
        {
            bool found = false;

            // Try the smallest possible character.
            for (int c = 0; c < 26; c++)
            {
                if (half[c] == 0)
                    continue;

                half[c]--;
                left += (char)('a' + c);

                // Fill the remaining positions with
                // the largest possible characters.
                string remaining = "";

                for (int x = 25; x >= 0; x--)
                {
                    if (half[x] > 0)
                        remaining += new string((char)('a' + x), half[x]);
                }

                string candidateLeft = left + remaining;
                string candidate = BuildPalindrome(candidateLeft);

                if (string.Compare(candidate, target, StringComparison.Ordinal) > 0)
                {
                    found = true;
                    break;
                }

                // Undo the choice.
                left = left.Substring(0, left.Length - 1);
                half[c]++;
            }

            if (!found)
                return "";
        }

        string answer = BuildPalindrome(left);

        return string.Compare(answer, target, StringComparison.Ordinal) > 0
            ? answer
            : "";
    }
}
